### DISCUSSION TITLE
Overriding result wheel tags

### DISCUSSION THREAD

**Discussion by MKuranowski** [mesonbuild/meson-python#580] [Q&A]
I'm piggy-backing on work done in #573 to enable cross-compilation with `zig cc`. It turns out pretty simple, and I have a working example on <https://github.com/MKuranowski/meson-python-zig/tree/942fb731b3e4b2fa84010e442e90ed14cac4b2e5>. The issue is that cross-compiled wheels still use the tags from the host systems - even though I'm compiling for Windows x86_64, the wheel uses a very specific `cp311-cp311-linux_x86_64` tag from the host machine.

```terminal
$ python -m build -w "-Csetup-args=--cross-file=$PWD/cross-x86_64-windows.ini"
...
+ /usr/bin/ninja
[2/2] Linking target libextern.dll
[4/4] /home/mikolaj/Documents/Code/meson-python-zig/pypkg/__main__.py
Successfully built pypkg-0.0.1-cp311-cp311-linux_x86_64.whl
```

The tag is too specific for the use case. Since I'm building a dynamic shared library, not dependent on Python, the Python and and ABI tags can be simply `py3` and `none`, however the most glaring issue is with the platform tag - the built wheel won't run on Linux.

The tags can be manually fixed with `wheel tags dist/pypkg-0.0.1-cp311-cp311-linux_x86_64.whl --python-tag py3 --abi-tag none --platform-tag win_amd64`, but running that at scale becomes problematic. [enscons has a way to override the wheel tags](https://dholth.github.io/enscons/#WHEEL_TAG), is there a similar possibility with meson-python?

**Comment by rgommers** [comment]
Thanks for sharing @MKuranowski! I'm not sure of the answer yet here, but I'm certainly interested in this use case and will try your code later.

> I'm piggy-backing on work done in #573

Wrong issue/PR number? That discussion is unrelated to cross compilation.

> but running that at scale becomes problematic. [enscons has a way to override the wheel tags](https://dholth.github.io/enscons/#WHEEL_TAG), is there a similar possibility with meson-python?

Such a thing doesn't exist right now in meson-python.

**Comment by MKuranowski** [comment]
> That discussion is unrelated to cross compilation.

Yes, I started from a simple shared library and tried to cross-compile it. I'm guessing cross-compiling extensions would be more useful, but my guess is that you'd run into linking issues (not sure, haven't tried).

> Such a thing doesn't exist right now in meson-python.

Thanks, I'll find a workaround to automatically call `wheel tags`.

**Reply by rgommers** [reply]
> I'm guessing cross-compiling extensions would be more useful, but my guess is that you'd run into linking issues

There are some issues with complex packages, but it does work in general. I've had reports of successful cross compilation of SciPy for example. 

The tracking issue for cross compilation support is gh-321.

I've played with your example repo some more, and I think it's feasible to get the wheel tags correct automatically:
- we can derive whether it's a cross build from whether `--cross-file` is present in `setup-args`,
- we can get the target platform from `build-dir/meson-info/intro-machines.json`,
- the logic for the `interpreter` and `abi` tags is mostly already correct, except:
    - it doesn't handle the case of "regular shared library with ctypes" yet (a shared library in `platlib` is interpreted as a Python extension module)
    - we need to get the target interpreter from Meson rather than assuming it's the same one used by `meson-python`. It looks like we can get it from `meson-info/intro-dependencies.json`. And it's missing if we don't call `py.dependency()` and don't build a Python extension module. Which would nicely deal with the `ctypes` situation as well.
